_posts/2020-12-01-the-road-to-the-jit.md

```diff
@@ -0,0 +1,372 @@
+
+> HiPE and execution of HiPE compiled code only have limited support
+  by the OTP team at Ericsson. The OTP team only does limited
+  maintenance of HiPE and does not actively develop HiPE. HiPE is
+  mainly supported by the HiPE team at Uppsala University.
+
+### Other Outcomes the HiPE Project
+
+I think it is fair to say that Erlang/OTP would look very different
+today if it hasn't been for the HiPE project.  Here are the major
+contributions from the HiPE project to OTP:
+
+* A new [staged tag scheme][newtags] in OTP R7.  The new tag scheme
+  allowed the Erlang system to address the full 4Gb address space (the
+  previous tag scheme only supported addressing the lower 1 Gb).
+  Surprisingly, the new tag scheme also improved performance.
+
+* The [Core Erlang][core] intermediate representation used in the
+  Erlang compiler to this day.  For more information, see
+  [An introduction to Core Erlang][core_introduction] and
+  [Core Erlang by Example][core_by_example].
+
+* [Dialyzer][dialyzer] (DIscrepancy AnaLYZer for ERlang programs),
+  started out as a type analysis pass for the HiPE native compiler,
+  but soon become a tool for Erlang programmers to help find bugs and
+  unreachable code in their applications.
+
+* [Bit strings and binary comprehensions][bitstrings].
+
+* Introducing [`try`...`catch`][exception] in OTP R10.
+
+* Repeatedly suggesting that Erlang needed a literal pool for premade
+  literal terms (instead of constructing them each time they are
+  used).  At one of our meetings between the HiPE team and the OTP
+  team, I remember [Richard Carlsson][richcarl] pointing out to me that it would
+  be nice for [Wings3D][wings3d] to have floating point literals.  The
+  OTP team implemented literal pools in OTP R12.
+
+[dialyzer]: http://erlang.org/doc/apps/dialyzer/index.html
+[newtags]: http://www.it.uu.se/research/publications/reports/2000-029/2000-029-nc.pdf
+[bitstrings]: http://user.it.uu.se/%7Epergu/papers/erlang05.pdf
+[core]: https://www.it.uu.se/research/group/hipe/cerl/doc/core_erlang-1.0.3.pdf
+[core_introduction]: http://www.erlang.se/workshop/carlsson.ps
+[core_by_example]: http://blog.erlang.org/core-erlang-by-example
+[exception]: https://erlang.org/workshop/2004/exception.pdf
+[richcarl]: https://github.com/richcarl
+[wings3d]: http://www.wings3d.com
+
+### The Tracing JIT projects
+
+There have been three separate research projects that tried to develop
+a tracing JIT for Erlang.
+
+A tracing JIT (Just In Time compiler) is a JIT that runs in two phases:
+
+* First it traces execution to find sequences of hot (frequently executed)
+  code.
+
+* It then rewrites the found traces to native code.
+
+The goals for the three JIT projects were:
+
+* The JIT should work automatically with no need for the user to identify
+  which modules to compile to native code beforehand.
+
+* There should be total feature compatibility with the non-JIT
+  BEAM. In particular, tracing, scheduling behavior, save calls, and
```

Comment:
Maybe add a link save calls here? Few people are familiar with it.

Reply:
Do we have a way to link directly to an option of `process_flag/2`? I didn't find any.

Reply:
I couldn't find any either, and it looks pretty difficult to add one that will produce stable links. Let's just leave it.

Reply:
I added a link to `process_flag/2`.